### ISSUE TITLE
feat(host): swap files + crash recovery for unsaved buffers

### ISSUE THREAD

**Issue by mxaddict** [kryptic-sh/hjkl#185]
## What it does

Auto-persist unsaved buffer state to a swap file on disk every few
seconds (and on idle). If hjkl crashes, panics, or is `kill -9`'d, the
next `:e <file>` detects the swap and offers to recover. Mirrors vim's
`.foo.swp` mechanism.

### Behavior

- On `:e <file>`, hjkl writes `~/.local/share/hjkl/swap/<hash>.swp` —
  a binary snapshot of the editor state (rope contents + cursor +
  marks + jump list). Hash is a stable derive of the canonical
  filesystem path so the same file always picks the same swap path.
- Swap is updated on:
  - Buffer modification + idle timer (default 4 seconds since last
    keystroke, vim's `updatetime`)
  - Explicit `:preserve` ex command
  - `<C-z>` / Suspend signal
- Swap is deleted on:
  - Successful `:w` (buffer-no-longer-dirty)
  - Buffer close without modifications
  - User confirms recovery and saves
- On `:e <file>` if a swap exists AND its mtime is newer than the
  file's mtime → recovery prompt:
  ```
  E325: Swap file ~/.local/share/hjkl/swap/<hash>.swp found
  Created: 2026-05-19 14:33:21 (8 minutes ago)
  Recover? [y/N/q]
  ```

### Recovery flow

- `y` → restore from swap, open the file, mark buffer dirty
- `N` → ignore swap, open file fresh, delete swap on next save
- `q` → quit (don't open the file)

### Edge cases

- Concurrent edit by another hjkl instance → swap exists + holds a
  PID lock. New instance sees the lock, reads the PID, checks if
  alive via `kill -0`. Alive → error "file already open in PID N";
  dead → swap-recovery flow.
- `:e` of a never-saved buffer (`hjkl` then `i…`): write swap at
  `~/.local/share/hjkl/swap/scratch_<uuid>.swp`. Recovery prompts
  to "save as" instead of restoring over the (non-existent) file.

## Why

Without this, any crash / OS kill / power loss = silent data loss
proportional to last save. Users assume modern editors have this
(VSCode auto-saves; vim has .swp; Neovim has undofile). hjkl
currently has zero crash safety net.

## Proposed scope

1. **Snapshot format** — reuse `EditorSnapshot` (already v5 per
   recent uppercase-marks work) + add file metadata header
   (canonical path, mtime, PID, write-time). `bincode` serialize.
2. **Swap-dir path** — XDG `STATE_HOME/hjkl/swap/` (use
   `hjkl-xdg`). Auto-create on first write.
3. **Hash derivation** — SHA-256 of canonical path → first 16 hex
   chars. Survives renames if user re-opens via the renamed path
   (treated as a new file, which is correct).
4. **Write trigger** — idle timer (`updatetime` config option,
   default 4000ms) fires on inactivity post-modification. Use the
   existing event loop's idle hook.
5. **Atomic write** — write to `.swp.tmp` + rename. Avoid partial
   swaps on crash mid-write.
6. **Recovery UI** — on `:e`, before loading file, check swap.
   Match swap header path to current path. If mtime newer, show
   recovery prompt via existing `hjkl-prompt` infra.
7. **PID lock** — append the writer PID to swap header. On open,
   check liveness with `kill -0 <pid>`; if alive AND it's another
   hjkl process, refuse to open with a clear error.
8. **`updatetime` ex option** — `:set updatetime=4000`. Configurable.
9. **`:preserve` ex command** — force-write swap immediately.
10. **`:recover [file]`** — explicit recovery for users who lost
    files to a crash.

## Acceptance

- Edit a file, don't save, send `kill -9` to hjkl. Re-open the
  file: recovery prompt offers to restore the unsaved edits.
- `:w` deletes the swap (no stale swap warning next open).
- Open the same file in two hjkl instances: second instance errors
  with "file already open in PID N" instead of silently letting both
  edit.
- Idle for 4+ seconds with a dirty buffer → swap file appears on
  disk and stays current.
- Process killed mid-swap-write doesn't leave a corrupt swap (atomic
  rename).

## Effort

**M** — snapshot serialize (already exists, +file header ~30 LOC),
swap-dir + path hashing (~50 LOC), idle-timer hook (~40 LOC), atomic
write (~30 LOC), recovery prompt (~80 LOC), PID lock (~50 LOC),
`:preserve` + `:recover` (~40 LOC), tests (~200 LOC + crash-recovery
integration test ~80 LOC). ~600 LOC.

## Non-goals

- Cross-session undo history (`'undofile`) — separate feature; share
  some plumbing but distinct UX
- Real-time autosave-to-disk (overwrites the actual file every N
  seconds). Vim explicitly doesn't do this and neither should we.
- Cloud-sync of swaps. No.

**Comment by mxaddict** [comment]
Core shipped in 438c5763:

- Swap format: postcard header (magic/version/canonical-path/mtimes/cursor) + raw rope-chunk body, atomic .swp.tmp→rename. Stored at `XDG_CACHE_HOME/hjkl/swap/<fnv16>.swp` (per maintainer call — cache not state).
- Idle writer on `updatetime` (option `ut`, default 4000ms), gated on dirty_gen so no busy-loop / disk hammer.
- `:w` deletes swap; `:preserve` force-writes.
- Recovery prompt on `:e` when swap newer than file (y restore / N disk / q abort); stale swaps auto-deleted.
- Huge-file-safe (rope.chunks() stream, no Vec<String> snapshot). Cross-platform (no unix-only calls) for 3-OS CI.

Deferred to follow-ups (TODO(#185) markers in swap.rs):
- [ ] PID-lock multi-instance detection (`kill -0` liveness — unix-specific, needs a portable design)
- [ ] `:recover [file]` explicit command
- [ ] scratch/never-saved-buffer swaps

Leaving open until those land.

**Comment by mxaddict** [comment]
Follow-up shipped in 90db5325:

- **PID-lock multi-instance**: swap header v2 carries writer_pid. `:e` of a file whose swap is owned by a LIVE other process is refused with E325 + the slot is removed (no half-open buffer). Liveness via `kill(pid,0)` on unix; non-unix falls back to no-lock (recovery still works, multi-instance refusal is unix-only for now).
- **`:recover [file]`**: explicit forced recovery, bypasses the mtime stale-delete gate. Empty arg = current buffer; path arg = `:e` then force-recover.

Still deferred (TODO(#185)):
- [ ] Scratch/never-saved-buffer swaps (needs a startup-scan + recovery-picker UX — separate feature)
- [ ] Windows multi-instance liveness (currently no-lock on non-unix)
- [ ] Delete swap on clean buffer-close of an unmodified buffer (today only `:w` deletes)

Leaving open for the scratch-buffer piece.

**Comment by mxaddict** [comment]
Core + all three sub-tasks shipped in v0.28.x: swap format, write-on-open + idle persistence, crash-recovery prompt (set_content restore), PID-lock multi-instance read-only, and scratch/never-saved buffer swaps with dead-writer orphan recovery. Two remainders — the dirty-buffer keep/reload/diff prompt and a Windows liveness check — are tracked in #241. Closing as core-complete.
